include/cuco/detail/open_addressing_impl.cuh

```diff
@@ -0,0 +1,534 @@
+/*
+ * Copyright (c) 2023, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cuco/detail/__config>
+#include <cuco/detail/common_functors.cuh>
+#include <cuco/detail/common_kernels.cuh>
+#include <cuco/detail/storage/counter_storage.cuh>
+#include <cuco/detail/tuning.cuh>
+#include <cuco/extent.cuh>
+#include <cuco/probing_scheme.cuh>
+#include <cuco/storage.cuh>
+#include <cuco/utility/traits.hpp>
+
+#include <thrust/iterator/constant_iterator.h>
+#include <thrust/iterator/counting_iterator.h>
+#include <thrust/iterator/transform_iterator.h>
+
+#include <cub/device/device_select.cuh>
+
+#include <cuda/atomic>
+
+namespace cuco {
+namespace experimental {
+namespace detail {
+/**
+ * @brief An open addressing impl class.
+ *
+ * @note This class should NOT be used directly.
+ *
+ * @throw If the size of the given slot type is larger than 8 bytes
+ * @throw If the given key type doesn't have unique object representations, i.e.,
+ * `cuco::bitwise_comparable_v<Key> == false`
+ * @throw If the probing scheme type is not inherited from `cuco::detail::probing_scheme_base`
+ *
+ * @tparam Key Type used for keys. Requires `cuco::is_bitwise_comparable_v<Key>`
+ * @tparam Value Type used for storage values.
+ * @tparam Extent Data structure size type
+ * @tparam Scope The scope in which operations will be performed by individual threads.
+ * @tparam KeyEqual Binary callable type used to compare two keys for equality
+ * @tparam ProbingScheme Probing scheme (see `include/cuco/probing_scheme.cuh` for choices)
+ * @tparam Allocator Type of allocator used for device storage
+ * @tparam Storage Slot window storage type
+ */
+template <class Key,
+          class Value,
+          class Extent,
+          cuda::thread_scope Scope,
+          class KeyEqual,
+          class ProbingScheme,
+          class Allocator,
+          class Storage>
+class open_addressing_impl {
+  static_assert(sizeof(Value) <= 16, "Container does not support slot types larger than 16 bytes.");
```

Comment:
Shall we restrict this to 8 bytes until we implement b2b CAS?

Reply:
Also, should we move this assert to the public API of map/set so that this restriction is visible to the user? Edit: The assert is only missing in the public `static_map` header